### ISSUE TITLE
doesn't run as standalone clone, README doesn't reflect current architecture

### ISSUE THREAD

**Issue by eflister** [trevordixon/LilyBin#80]
hi -
i'm forking in order to swap in my own parser for a custom input format that compiles to lilypond.  the current README hasn't been updated to reflect the move to aws lambda, so that threw me, and it took me a while to realize the dependency on https://github.com/LilyBin/lambda-compile.  

it would be nice for new clones to be able to run standalone out of the box, without requiring setting up/authenticating to cloud services -- perhaps some way to serve from https://github.com/LilyBin/lambda-compile/blob/master/test/local.js?

i haven't done cloud stuff before, but i can see you've done nice work here and i'm trying to figure out the best way to build on what you've done.  if it's possible for my server to use your lambda instance for running lilypond, that would be ideal.  would you be willing to provide me with the necessary authorization?  otherwise, i can try to set up my own instance, since the free tier looks more than adequate -- this is just a hobby project that won't have any serious traffic.  S3, on the other hand, looks like it will start charging after 1 year -- could you also authorize using your storage, or alternatively we can look at re-enabling an option for local storage from your previous versions alongside the cloud option?

thanks for help and great work on lilybin!

**Comment by trevordixon** [comment]
Feel free to hit the web service however you want! Example:

    curl -X POST -H "Content-Type:application/json" https://7icpm9qr6a.execute-api.us-west-2.amazonaws.com/prod/prepare_preview/stable -d '{"code": "% LilyBin\n\\score{\n\t{\n\t\t\\repeat unfold 120 { c4. d e f }\n\t}\n\n\t\\layout{}\n\t\\midi{}\n}\n"}'

**Comment by eflister** [comment]
ha ha, you're right, thanks, my server can POST to the lambda instance fine.  what i can't see are the generated files on s3, so i was assuming the same applied to all aws services.  so now the problem is:
`XMLHttpRequest cannot load https://s3-us-west-2.amazonaws.com/lilybin-scores/1492672820446-4974-843996131-6xf7pflb1oewb3xr.pdf?t=1492672821897. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:3001' is therefore not allowed access.`

**Comment by trevordixon** [comment]
Try again. I updated the configuration to provide CORS headers for any origin, so it should work now.

**Comment by eflister** [comment]
it's working great, thanks!  :)  pls don't forget to update the README...
also, any suggestions for allowing an offline mode?  it would be helpful to have stand-ins for the aws services when there's no network.  i'm guessing there's no easy way, but thought i'd ask.  :)

**Comment by trevordixon** [comment]
Nope. Maybe LilyPond will run in a browser someday, but not today!

**Comment by eflister** [comment]
i mean a local substitute (that would call a local lilypond) of aws services that the node server could hit.  it would run off of the lambda function that `lambda-compile` generates...

**Comment by trevordixon** [comment]
Nothing like that exists at the moment. The easiest way to do it might be to use something like https://github.com/ashiina/lambda-local with https://github.com/LilyBin/lambda-compile?

**Comment by eflister** [comment]
thanks for the link, that looks exactly like what i had in mind!

**Comment by erantapaa** [comment]
Hey - the web service is great! But how do I access the resulting files?

I've got a response like:

```
{"stdout":"","stderr":"Processing `input.ly'\nParsing...\ninput.ly:1: warning: no \\version statement
 found, please add\n\n\\version \"2.18.2\"\n\nfor future compatibility\nInterpreting music...[8][16][24]
[32][40][48][56]\nPreprocessing graphical objects...\nInterpreting music...\nMIDI output to
 `rendered.midi'...\nFinding the ideal number of pages...\nFitting music on 1 or 2 pages...\nDrawing
 systems...\nLayout output to `rendered.ps'...\nConverting to `./rendered.pdf'...\nSuccess:
 compilation successfully completed\n","id":"1512605968000-8124-967311653-x6csfkvtk5p3z0k9"
,"files":{"pdf":true,"midi":true}}
```

Looks like the PDF and MIDI products are associated with id `1512605968000-8124-967311653-x6csfkvtk5p3z0k9`.

What's the next step?

UPDATE: Just realized it was a snap to get the UI running locally, so I'll dive into the code to see if I can figure it out.

**Comment by trevordixon** [comment]
Looking at the network tab in Chrome dev tools, you can see it's loading URLs like this: https://s3-us-west-2.amazonaws.com/lilybin-scores/1512636370669-10664-740456891-pu2wq293vcuwhfr.pdf

So substitute the id into a URL like this: https://s3-us-west-2.amazonaws.com/lilybin-scores/{id}.pdf

**Comment by erantapaa** [comment]
Thanks!

I hacked the code so you could run it from a static page server instead of its nodejs server. The main change I had to make was to replace the `$.get` call [(here)][1] (since there isn't a server anymore.)

I hardcoded `data.version` to `"stable"` and `data.code` to the default sample code and it appears that everything except for saving worked.

Being able to host the app on a static page server (like github user pages) is really convenient. Perhaps it could be a config option?

Oh - yeah - just remembered that the other thing I had to change were the absolute URLs to the local resources. But once I fixed those the app just worked.

[1]: https://github.com/LilyBin/LilyBin/blob/1df09dc1d83707d79fb91a6c77ce27ae819a3680/htdocs/js/main.js#L207

**Comment by erantapaa** [comment]
Here is an example of lilybin running on static page server:

http://erantapaa.github.io/daily-programmer/343I/htdocs/index.html

When you open the application it generates a random composition according to
the rules of "Mozart's Musical Dice Game". The input field in the nav bar selects the
random seed. A seed value of 0 shows all of the 176 measures which could be used
in the composition.

More info about the game may be found [here][1].

[1]: https://www.reddit.com/r/dailyprogrammer/comments/7i1ib1/20171206_challenge_343_intermediate_mozarts/
